server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -126,7 +128,17 @@ private Map<String, UploadedSegmentMetadata> readLatestMetadataFile() throws IOE
 
     private Map<String, UploadedSegmentMetadata> readMetadataFile(String metadataFilename) throws IOException {
         try (IndexInput indexInput = remoteMetadataDirectory.openInput(metadataFilename, IOContext.DEFAULT)) {
-            Map<String, String> segmentMetadata = indexInput.readMapOfStrings();
+            // BufferedChecksumIndexInput reader keeps computing checksum for all the bytes we read from indexInput
+            // This checksum is then compared with checksum in footer
+            ChecksumIndexInput in = new BufferedChecksumIndexInput(indexInput);
+            CodecUtil.checkHeader(
+                in,
+                UploadedSegmentMetadata.METADATA_CODEC,
+                UploadedSegmentMetadata.CURRENT_VERSION,
```

Comment:
We can support a min version of metadata file for each OS version and deprecate old versions as we go along. When checking header we get in return the version of the metadata file we read. Which we can use to decide how to read the metadata file.

But i feel it won't this simple in the long run. Currently, migrating indexes to new OS version is restricted due to how lucene restricts upgrades, adding more blockers/conditions to this might not be ideal. Ideally we should be able to automatically upgrade the metadata files to new version when read if the metadata version is old. We could follow the same approach as lucene does, by allowing only one major version jump at a time? 

This could be part of recovery from remote store flow? I was thinking if we could automatically upgrade the metadata files to new versions whenever possible during recovery flow and try to avoid changes to metadata file where such automatic upgrade is not possible.

I think we can spend some time to think about it and use a uniform version evolution strategy for both translog and segment metadata files. do you think we should create a separate issue for this? or include in this PR?

Reply:
Created a separate issue for handling backward compatibility https://github.com/opensearch-project/OpenSearch/issues/6123